src/Protocol.cpp

```diff
@@ -47,6 +48,28 @@ int protocol::extractPacket(uint8_t const* buffer, size_t size,
     return 2 + payload_end - buffer;
 }
 
+std::pair<uint8_t const*, uint8_t const*> protocol::getPayload(uint8_t const* buffer) {
+    auto parsed_length = parseLength(buffer + 2);
+    return make_pair(parsed_length.second, parsed_length.first + parsed_length.second);
+}
+
+uint8_t* protocol::encodeFrame(uint8_t* buffer,
+                               uint8_t const* payload_begin,
+                               uint8_t const* payload_end) {
+    buffer[0] = SYNC_0;
+    buffer[1] = SYNC_1;
+
+    size_t payload_length = payload_end - payload_begin;
+    uint8_t* length_end = encodeLength(buffer + 2, payload_length);
+    std::memcpy(length_end, payload_begin, payload_length);
```

Comment:
I think you should be checking the buffer size to make sure the payload fits here

Reply:
Agreed. I went for the "the caller should know what he's doing", but that really is a micro-optimization with possibly catastrophic results. I'll check if there are other places where some sanity checks could be put.